tlcc/Makefile

```diff
@@ -6,15 +6,25 @@ TOP = ..
 include $(TOP)/build.mk
 
 LD_LIBRARY_PATH := LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:$$(pwd)/enclave/lib
-SRC := tlcc.go
-OUTPUT := tlcc.so
 
-$(OUTPUT): $(SRC)
-	$(LD_LIBRARY_PATH) $(GO) build -o $@ -buildmode=plugin $<
+FABRIC_TLCC = $(FABRIC_PATH)/core/scc/tlcc
+FABRIC_PEER_BIN = $(FABRIC_PATH)/build/bin/peer
 
-build: $(OUTPUT)
+build: fabric-peer
 
-test: build
+fabric-peer: $(FABRIC_PEER_BIN)
```

Comment:
This does not belong here. Any reason to have the peer built in our build process?
Besides, the peer patch is documented but not applied here -- it's applied in travis and docker.

I see that it could be convenient to have this step in the build process. In this case it should probably be an `fpc-peer` target in the top-level makefile, which also applies all the patches we have -- and we should remove this step in travis and docker.

Additionally, to make sure the peer is right, we might add a capability check in tlcc and tests to test whether a query to tlcc fails -- oth. it will badly fail later when it does not find the system chaincode.

Reply:
I can go either way on this.

On a principal of owning what we're doing, we should build the peer as we modify it with a patch.

On the other hand, someone running this does need to know to manually swap the peer binary out.

So by automating this step are we fooling later users into skipping a step? It is something which affirmatively needs to be done manually.

Until there are outside users, I think automating the process is helpful, wherever it occurs.

Reply:
I am fine with calling it fpc-peer.

At least we need this target for us to build the peer for our integration tests. If you give this to the end users we can discuss. Would it make more sense to put the patched fpc peer binary somewhere in the fpc folder?! $FPC_PATH/fabric/build/bin ??

Reply:
The target name is important but secondary.
Yes, we need it for the integration tests, because the we run those in the concept release.

The main points are "where" it is and "what" it does.
The travis file answers to this https://github.com/hyperledger-labs/fabric-private-chaincode/blob/169cfe756de05d1495aef8175a831cff149744fe/.travis.yml#L67.
What we do in travis is to: (1) apply the patches, (2) recompile the peer, (3) run make (with the integration tests).
So if you create an `fpc-peer` target in the top-level makefile:
* you can move (1) and (2) from travis to the new target
* you automate the procedure for the end user, because now the user still has to apply the patches manually whenever he starts from scratch
* you can make sure you build the "patched" peer  -- in the current form you might build the unpatched peer and only discover later that tlcc is not there.

Reply:
Sounds good to me

Reply:
I created this peer target in `fabric/Makefile`. It patches with peer with all patches we have and compiles it; moreover, we have a target to build the other fabric binaries (orderer, confitxgen ....;

Reply:
> ok, did now some quick tests, but while ledger_init did work, the commit of helloworld failed with
> 
> ```
> 20-04-09 14:05:28.115 PDT [grpc] createTransport -> DEBU 025 grpc: addrConn.createTransport failed t
> o connect to {127.0.0.1:7051 0  <nil>}. Err :connection error: desc = "transport: error while dialing:
>  dial tcp 127.0.0.1:7051: connect: connection refused". Reconnecting...
> 2020-04-09 14:05:28.116 PDT [grpc] HandleSubConnStateChange -> DEBU 026 pickfirstBalancer: HandleSubCo
> nnStateChange: 0xc000453340, TRANSIENT_FAILURE
> Error: error getting endorser client for query: endorser client failed to connect to 127.0.0.1:7051: f
> ailed to create new connection: connection error: desc = "transport: error while dialing: dial tcp 127
> .0.0.1:7051: connect: connection refused"
> + die 'test failed: /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledge
> r/fabric/build/bin/peer chaincode query -o localhost:7050 -C mychannel -n hello -c {"Args":["__getEncl
> avePk"]}'
> ++ basename ./fabric/bin/peer.sh
> + recho 'peer.sh: test failed: /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com
> /hyperledger/fabric/build/bin/peer chaincode query -o localhost:7050 -C mychannel -n hello -c {"Args":
> ["__getEnclavePk"]}'
> + echo 'peer.sh:' test failed: /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com
> /hyperledger/fabric/build/bin/peer chaincode query -o localhost:7050 -C mychannel -n hello -c '{"Args"
> :["__getEnclavePk"]}'
> peer.sh: test failed: /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperled
> ger/fabric/build/bin/peer chaincode query -o localhost:7050 -C mychannel -n hello -c {"Args":["__getEn
> clavePk"]}
> + exit 111
> ```
> 
> and /tmp/hyperledger/test/peer.err shows a tlcc crash
> 
> ```
> rusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/tlcc_enclave/enclave/ledger.cpp:560) [Enclave] Ledger:                     \-> key = _lifecycle.namespaces/fields/hello/Sequence version = blockNum: 5; txNum: 0
> 2020-04-09 14:05:28.035 PDT [tl-enclave] golog -> INFO 2e8d DEBUG  (/home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/tlcc_enclave/enclave/ledger.cpp:560) [Enclave] Ledger:                     \-> key = hello.MRENCLAVE version = blockNum: 0; txNum: 0
> 2020-04-09 14:05:28.035 PDT [tl-enclave] golog -> INFO 2e8e DEBUG  (/home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/tlcc_enclave/enclave/ledger.cpp:582) [Enclave] Ledger:             \-> Writes:
> 2020-04-09 14:05:28.035 PDT [tl-enclave] golog -> INFO 2e8f ERROR  (/home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/tlcc_enclave/enclave/ledger.cpp:625) [Enclave] Ledger: >>> NO MRENCLAVE found for: ^T.MRENCLAVE
> fatal: morestack on g0
> SIGTRAP: trace trap
> PC=0xae1872 m=4 sigcode=128
> 
> goroutine 0 [idle]:
> runtime.abort()
>         /home/msteiner/go/go1.14.linux-amd64/src/runtime/asm_amd64.s:859 +0x2
> runtime.morestack()
>         /home/msteiner/go/go1.14.linux-amd64/src/runtime/asm_amd64.s:416 +0x25
> 
> goroutine 175 [syscall]:
> runtime.cgocall(0x1576c80, 0xc0029e7f28, 0x17ed340)
>         /home/msteiner/go/go1.14.linux-amd64/src/runtime/cgocall.go:133 +0x5b fp=0xc0029e7ef8 sp=0xc0029e7ec0 pc=0xa7d03b
> github.com/hyperledger-labs/fabric-private-chaincode/tlcc/enclave._C2func_tlcc_send_block(0x1bd500000002, 0x7f48380154d0, 0x1227, 0x7f4800000000, 0x0, 0x0)
>         _cgo_gotypes.go:86 +0x55 fp=0xc0029e7f28 sp=0xc0029e7ef8 pc=0x1547615
> github.com/hyperledger-labs/fabric-private-chaincode/tlcc/enclave.(*StubImpl).NextBlock(0xc0002e0c08, 0xc002e6c000, 0x1227, 0x1227, 0x0, 0x0)
>         /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger-labs/fabric-private-chaincode/tlcc/enclave/enclave_stub.go:132 +0xa2 fp=0xc0029e7f80 sp=0xc0029e7f28 pc=0x15487e2
> github.com/hyperledger/fabric/core/scc/tlcc.(*TrustedLedgerCC).readBlocks(0xc0003bb500, 0x1b4ab60, 0xc0003baea0)
>         /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger/fabric/core/scc/tlcc/tlcc.go:148 +0xa7 fp=0xc0029e7fc8 sp=0xc0029e7f80 pc=0x154b787
> runtime.goexit()
>         /home/msteiner/go/go1.14.linux-amd64/src/runtime/asm_amd64.s:1373 +0x1 fp=0xc0029e7fd0 sp=0xc0029e7fc8 pc=0xae1f61
> created by github.com/hyperledger/fabric/core/scc/tlcc.(*TrustedLedgerCC).joinChannel
>         /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger/fabric/core/scc/tlcc/tlcc.go:133 +0x28c
> 
> goroutine 1 [chan receive, 49 minutes]:
> github.com/hyperledger/fabric/internal/peer/node.serve(0x277e610, 0x0, 0x0, 0x0, 0x0)
>         /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger/fabric/internal/peer/node/start.go:863 +0x4aa4
> github.com/hyperledger/fabric/internal/peer/node.glob..func6(0x264c1c0, 0x277e610, 0x0, 0x0, 0x0, 0x0)
>         /home/msteiner/src/poet-et-al/go-trusted-chain-code.v2.0/src/github.com/hyperledger/fabric/in
> ```

now we integration tests enabled, travis also produces this error.